### PR TITLE
Kulagin Aleksandr / lab 2 / var 2

### PR DESCRIPTION
**Pull request by mahbhlddnhakkh**
Let's call function that returns void and has no parameters "empty function". Let's say empty function has n blocks. In order to put this empty function in a place it's called, we split called function, create n blocks in calling function and create association list, which associate each created block with it's corresponding block from empty function. Next we copy all instructions from associated blocks. Also we change branch instructions so they send to correct block in called function. And we replace return statements to branch instruction that sends to block with instructions after empty function call.

test.ll file contains 8 tests. Each test contains: C code, plugin's IR output for reference, checks and IR code for checks.

Now let's talk about problems with my code:
1. Because I have to reorganize variable and block names, their names no longer match with names from original IR code. So testing is a bit complicated. And I'm not even sure if I'm testing all of this correctly.
2. The splitting function names block ".split" instead of a number which doesn't look pretty, but I don't know how to fix it easily. It does not affect the plugin because it's just a name.

### CODE REVIEW

File: llvm/labs/CMakeLists.txt [hunk 1]
```diff
@@ -7,4 +7,4 @@ ENDFOREACH()
 
 set(LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS} PARENT_SCOPE)
 
-message("LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS}")
+message("LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS}")
```
Comment (m-ly4) [inline llvm/labs/CMakeLists.txt]: Please revert this change as not necessary.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+namespace {
+llvm::BasicBlock *splitBlockBefore(llvm::BasicBlock *Old,
+                                   llvm::Instruction *SplitPt,
+                                   llvm::DomTreeUpdater *DTU = nullptr,
+                                   llvm::LoopInfo *LI = nullptr,
+                                   llvm::MemorySSAUpdater *MSSAU = nullptr,
+                                   const llvm::Twine &BBName = "",
+                                   bool Before = true) {
+  return llvm::SplitBlock(Old, SplitPt, DTU, LI, MSSAU, BBName, Before);
```
Comment (m-ly4) [inline llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp:25]: This function creates basic block with name `.split` if BBName is not specified.
This is fine because basic block's name doesn't affect program's functionality.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+class KulaginMagicInlinePass
+    : public llvm::PassInfoMixin<KulaginMagicInlinePass> {
+public:
+  llvm::PreservedAnalyses run(llvm::Function &Func,
+                              llvm::FunctionAnalysisManager &) {
+    std::vector<llvm::Instruction *> InstrDeleteList;
+    llvm::LLVMContext &CTX = Func.getContext();
+    if (Func.getReturnType()->isVoidTy() &&
+        Func.getFunctionType()->getNumParams() == 0) {
+      EmptyFuncs.push_back(&Func);
```
Comment (m-ly4) [inline llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp]: `Empty...` word in variables confuses because I think that those are functions without any instruction. I'd suggest to name them as FuncsToInline or smth like this.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      }
+    }
+    for (llvm::Instruction *Instr : InstrDeleteList) {
+      Instr->eraseFromParent();
+    }
+    return llvm::PreservedAnalyses::all();
+  }
+
+private:
+  std::vector<llvm::Function *> EmptyFuncs;
```
Comment (m-ly4) [inline llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp]: I'd suggest to simplify a bit the pass in the following way:
It just searches for `call` instructions in a processed function and if called function doesn't have arguments and doesn't return anything then inline it after `call`?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/labs/CMakeLists.txt (hunk 1 above)
Comment (mahbhlddnhakkh) [reply llvm/labs/CMakeLists.txt]: > Please revert this change as not necessary.

I'm sorry I don't know how. But it doesn't change anything so why not just leave it like that?

**Comment by mahbhlddnhakkh** [comment]
I simplified the pass. So there's no need to collect functions to inline and use search. It just inlines instantly (if the function doesn't return anything and has no parameters of course). But it changes the behavior of circular calls from test8.ll . foo() has inline bar() and bar() has inline foo(), but they never go deeper than 1 level. Previously foo() hadn't inline bar() (it was just calling it) and bar() had 1 level foo() inline. It had a bit shorter IR code. The rest of the tests are not affected.

File: llvm/labs/CMakeLists.txt (hunk 1 above)
Comment (m-ly4) [reply llvm/labs/CMakeLists.txt]:
```suggestion
message("LLVM_TEST_DEPENDS ${LLVM_TEST_DEPENDS}")

```

Comment (m-ly4) [reply llvm/labs/CMakeLists.txt]: > > Please revert this change as not necessary.
> 
> I'm sorry I don't know how. But it doesn't change anything so why not just leave it like that?

Seems that your changes removed `\n` char in the end of last line. Actually, it is required to add new line symbol to the last line of any file in the LLVM project.
Regarding, accidental changes if they even don't affect anything it may complicate analysis of changes in a project if there are mess of non-related changes in a file.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    : public llvm::PassInfoMixin<KulaginMagicInlinePass> {
+public:
+  llvm::PreservedAnalyses run(llvm::Function &Func,
+                              llvm::FunctionAnalysisManager &) {
+    std::vector<llvm::Instruction *> InstrDeleteList;
+    llvm::LLVMContext &CTX = Func.getContext();
+    for (llvm::BasicBlock &Block : Func) {
+      for (llvm::Instruction &Instr : Block) {
+        if (llvm::CallInst *CallInstruction =
+                llvm::dyn_cast<llvm::CallInst>(&Instr)) {
```
Comment (m-ly4) [inline llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp]: In LLVM projects earlier check and exit from a loop or other compound statement is usually used to reduce blocks nesting in a code.
```suggestion
        auto *CallInstruction = llvm::dyn_cast<llvm::CallInst>(&Instr);
        if (!CallInstruction) {
          continue;
        }
```

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+          llvm::ValueToValueMapTy VMap;
+          llvm::BasicBlock *SplittedBlock = splitBlockBefore(&Block, &Instr);
+          llvm::BasicBlock *NextBlockPlace = SplittedBlock->getNextNode();
+          for (llvm::BasicBlock &ToInlineBasicBlock : *Called) {
+            llvm::BasicBlock *TmpBlock =
+                llvm::BasicBlock::Create(CTX, "", &Func, NextBlockPlace);
+            BlockMap[&ToInlineBasicBlock] = TmpBlock;
+            NextBlockPlace = TmpBlock->getNextNode();
+          }
+          for (llvm::BasicBlock &ToInlineBasicBlock : *Called) {
```
Comment (m-ly4) [inline llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp]: May these loops be joint? I mean we split a block in a processed function and loop through blocks and instructions of inlined functions inserting them in place?

File: llvm/labs/CMakeLists.txt (hunk 1 above)
Comment (mahbhlddnhakkh) [reply llvm/labs/CMakeLists.txt]: > > > Please revert this change as not necessary.
> > 
> > 
> > I'm sorry I don't know how. But it doesn't change anything so why not just leave it like that?
> 
> Seems that your changes removed `\n` char in the end of last line. Actually, it is required to add new line symbol to the last line of any file in the LLVM project. Regarding, accidental changes if they even don't affect anything it may complicate analysis of changes in a project if there are mess of non-related changes in a file.

Does that mean that I should add extra `\n` on 3 files that has been created by me (and start github actions again)? Or I just should be careful about it on the next labs? Because I didn't put extra `\n` in the first lab's files.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+              0,
+              llvm::dyn_cast<llvm::Value>(BlockMap[&Called->getEntryBlock()]));
+          InstrDeleteList.push_back(&Instr);
+        }
+      }
+    }
+    for (llvm::Instruction *Instr : InstrDeleteList) {
+      Instr->eraseFromParent();
+    }
+    return llvm::PreservedAnalyses::all();
```
Comment (m-ly4) [inline llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp:98]: Are you sure that results of all analyses are not staled after such a code transformation?

**Review by m-ly4 — CHANGES_REQUESTED** [review]

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp (hunk 6 above)
Comment (mahbhlddnhakkh) [reply llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp]: > May these loops be joint? I mean we split a block in a processed function and loop through blocks and instructions of inlined functions inserting them in place?

No. Because some branch instructions can call a block that was not created and associated yet. Basically any loop and if statement. The branch instruction there calls the next block.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp (hunk 7 above)
Comment (mahbhlddnhakkh) [reply llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp:98]: > Are you sure that results of all analyses are not staled after such a code transformation?

It passes circular call (tested in test8.ll) and I haven't came up with the situation where this pass fails, so I think yes. For each run it doesn't change any other functions other than the one that's presented on this run.

Comment (m-ly4) [reply llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp:98]: Analysis passes results are used by not only this pass, but in case of pipelines of passes they may be used by any further pass after yours one. If you don't reset analysis results, other passes may start to use all function structure information and their result may even corrupt valid behavior of compiled program.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp (hunk 6 above)
Comment (m-ly4) [reply llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp]: Ok, good.

File: llvm/labs/CMakeLists.txt (hunk 1 above)
Comment (m-ly4) [reply llvm/labs/CMakeLists.txt]: Yes, it would be nice of you to add a new line in the end of all your files in the code.

File: llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp (hunk 7 above)
Comment (mahbhlddnhakkh) [reply llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp:98]: > Analysis passes results are used by not only this pass, but in case of pipelines of passes they may be used by any further pass after yours one. If you don't reset analysis results, other passes may start to use all function structure information and their result may even corrupt valid behavior of compiled program.

I don't really know. It generates the correct IR code, the naming is correct, the dependency is also correct (names do not cross in a single function). I don't see a reason why any other pass might fail. Is it because I wrap inline function between 2 non-conditional branch instructions so that inline function will always be in a separate block?

Comment (m-ly4) [reply llvm/labs/lab2/kulagin_aleksandr/magic_inline.cpp:98]: This is not about current pas, this is about cached results of other analysis passes https://llvm.org/docs/Passes.html.
Just search for preserved analyses info.

**Review by m-ly4 — APPROVED** [review]

**Comment by mahbhlddnhakkh** [comment]
I have added the needed lines for other passes to work in a pipeline. So that when the transformation happens (every time we see a void function without parameters call), we don't preserve the analysis.

**Review by m-ly4 — APPROVED** [review]

**Comment by m-ly4** [comment]
Well done, thank you.
